File: Data Structures & Algorithms/island-perimeter/submission-0.py
class Solution:
    dirs = [[1,0],[-1,0],[0,1],[0,-1]]
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        
        rows = len(grid)
        cols = len(grid[0])
        for r in range (rows):
            for c in range (cols):
                if grid[r][c] == 1:
                    return self.dfs(grid,r,c)
        return 0
    
    def dfs(self,grid,r,c):
        
        rows = len(grid)
        cols = len(grid[0])

        if(r<0 or r>=rows or c<0 or c>=cols):
            return 1
        
        if grid[r][c] == 0 :
            return 1
        if grid[r][c] == -1:
            return 0
        grid[r][c] = -1
        
        perimeter = 0
        for dr,dc in self.dirs:
            perimeter+=self.dfs(grid,r+dr,c+dc)
        
        return perimeter

File: Data Structures & Algorithms/max-area-of-island/submission-0.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();
        int maxArea = 0;

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == 1) {
                    int area = dfs(grid, r, c);
                    maxArea = max(maxArea, area);
                }
            }
        }

        return maxArea;
    }

    int dfs(vector<vector<int>>& grid, int r, int c) {
        if (r < 0 || c < 0 || r >= grid.size() || c >= grid[0].size() || grid[r][c] == 0) {
            return 0;
        }

        grid[r][c] = 0; // Mark as visited
        int area = 1;

        int directions[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};
        for (int i = 0; i < 4; i++) {
            area += dfs(grid, r + directions[i][0], c + directions[i][1]);
        }

        return area;
    }
};
